benchrun/python/benchrun/_benchmark.py

```diff
@@ -0,0 +1,263 @@
+import abc
+import copy
+import gc
+import multiprocessing as mp
+import time
+import uuid
+import warnings
+from typing import Any, Dict, List
+
+from benchadapt import BenchmarkResult
+
+from .cache import CacheManager
+from .case import CaseList
+
+
+class Iteration(abc.ABC):
+    """
+    Abstract class defining how to run one iteration of a benchmark.
+
+    Attributes
+    ----------
+
+    name : str
+        A name for the benchmark. Should be specified when inheriting.
+    cache : CacheManager
+        An CacheManager instance for clearing the disk cache when specified.
+        Do not mess with this.
+    """
+
+    name: str
+    cache = None
+
+    def __init__(self) -> None:
+        assert self.name
+        self.cache = CacheManager()
+
+    def setup(self, case: dict) -> dict:
+        """
+        Code to run in each iteration before timing.
+
+        Parameters
+        ----------
+        case : dict
+            A dict where keys are parameters and values are scalar arguments for a benchmark
+
+        Returns
+        -------
+        A dict passed to the ``setup_results`` param of ``run()`` when the class is called.
+        """
+        return {}
```

Comment:
Hm, so I wrote this so it runs each iteration in a new process, so a global setup/teardown that runs in the main process would be less useful (still good for writing files, etc; less good for importing libraries and setting up contexts) and more confusing. Reading through arrowbench, I'm pretty sure it's running all iterations in a single new process. We could switch to that, if we want? It's maybe less good from a benchmarking perspective, but way simpler from a user perspective.

Reply:
Where does the timing happen? Inside the subprocess?

Reply:
Yep, that happens in the subprocess either way

Reply:
It depends on the use-case. If the setup block is "download this dataset", that can be global regardless; if it's "read in this dataset to time writing", yeah, it'd have to happen in each subprocess, whether that's per benchmark or per iteration. Subprocess per-benchmark is definitely faster; the only reason to keep per-iteration is because we think the isolation would result in lower-variance timings. It may prevent some caching/memoisation/JIT effects (maybe?? potentially, but there's no guaranteed effect size here), but doing a big test here would require implementing a lot of benchmarks.

Reply:
Indeed. It's possible we will want | need both approaches. IMO we shouldn't decide which until we do the test and have data to back up our recommendations. In the mean time, let's flag our uncertainty, pick one, merge it, and create the follow on to do it right in the future

Reply:
Yeah, actually it might not be all that hard to make two subclasses and allow both. And for the simplicity of programming, subprocess-per-benchmark is probably the default, which means I'll have to write most of both anyway if we want to just do it now.

Reply:
Alright, in separating out `setup()` and `teardown()` from `before_each()` and `after_each()`, it became a lot easier to just run it exactly like arrowbench, which seems good for continuity. We can reimplement subprocess-per-iteration if we want, but it will be a tad more finicky to program.

I also moved all the `*_results` params to an `env` dict attribute that stages can modify to store/pass state; that felt less messy and more like arrowbench anyway.